### PR TITLE
Add Caching to CBA_fnc_compatibleItems (Cleanup)

### PR DESCRIPTION
**Pull request by PabstMirror**
Added fixes and test case to @jokoho48 's #261

### CODE REVIEW

**Comment by jokoho48** [comment]
from my side this is good to go :+1: but @robalo is the maintainer

**Comment by robalo** [comment]
I have to catch up with things, don't wait for me. as long as you don't break the function... go for it.
What does this change do exactly ?

**Comment by esteldunedain** [comment]
> should we be concerned about polluting the namespace with a lot of variables ?

No, don't worry. This PR stores the cache on a dedicated namespace (location).

File: addons/jr/fnc_compatibleItems.sqf [hunk 1]
```diff
@@ -1,10 +1,11 @@
 /*
     Author: Karel Moricky
     Enhanced by Robalo
+    adding Caching by joko // Jonas
```
Comment (Killswitch00) [inline addons/jr/fnc_compatibleItems.sqf]: "Caching added by..."

File: addons/jr/fnc_compatibleItems.sqf [hunk 2]
```diff
@@ -15,32 +16,46 @@
...
-                } foreach configproperties [_cfgCompatibleItems, "isNumber _x"];
+                    _compatibleItems pushBackUnique _x;
+                    nil
+                } count (getArray _cfgCompatibleItems);
+            } else {
+                if (isClass _cfgCompatibleItems) then {
+                    {
+                        if ((getnumber _x > 0)) then {_compatibleItems pushBackUnique (configName _x)};
+                        nil
+                    } count configproperties [_cfgCompatibleItems, "isNumber _x"];
```
Comment (Killswitch00) [inline addons/jr/fnc_compatibleItems.sqf]: `configProperties`

File: addons/jr/fnc_compatibleItems.sqf (hunk 1 above)
Comment (jokoho48) [inline addons/jr/fnc_compatibleItems.sqf]: :D ok will fix it

File: addons/jr/fnc_compatibleItems.sqf [hunk 3]
```diff
@@ -15,32 +16,46 @@
...
                 {
-                    if (getnumber _x > 0 && {!((configname _x) in _compatibleItems)}) then {_compatibleItems pushBack configname _x};
-                } foreach configproperties [_cfgCompatibleItems, "isNumber _x"];
+                    _compatibleItems pushBackUnique _x;
+                    nil
+                } count (getArray _cfgCompatibleItems);
+            } else {
+                if (isClass _cfgCompatibleItems) then {
+                    {
+                        if ((getnumber _x > 0)) then {_compatibleItems pushBackUnique (configName _x)};
```
Comment (Killswitch00) [inline addons/jr/fnc_compatibleItems.sqf]: `getNumber`

File: addons/jr/fnc_compatibleItems.sqf [hunk 4]
```diff
@@ -15,32 +16,46 @@
...
-        } else {
-            if (isclass _cfgCompatibleItems) then {
+if (isNil QGVAR(namespace)) then {
+    GVAR(namespace) = call CBA_fnc_createNamespace;
+};
+private _compatibleItems = GVAR(namespace) getVariable _weapon;
+
+if (isNil "_compatibleItems") then {
+    _compatibleItems = [];
+    private _cfgWeapon = configfile >> "cfgweapons" >> _weapon;
```
Comment (Killswitch00) [inline addons/jr/fnc_compatibleItems.sqf]: `configFile`

File: addons/jr/fnc_compatibleItems.sqf [hunk 5]
```diff
@@ -15,32 +16,46 @@
...
+                        if ((getnumber _x > 0)) then {_compatibleItems pushBackUnique (configName _x)};
+                        nil
+                    } count configproperties [_cfgCompatibleItems, "isNumber _x"];
+                };
             };
-        };
-    } foreach configproperties [_cfgWeapon >> "WeaponSlotsInfo","isclass _x"];
-    if (_typefilter == 0) then {_compatibleItems} else {[_compatibleItems, {_typefilter == getnumber(configfile>>"cfgweapons">>_x>>"itemInfo">>"type")}] call BIS_fnc_conditionalSelect};
+            nil
+        } count configproperties [_cfgWeapon >> "WeaponSlotsInfo","isclass _x"];
```
Comment (Killswitch00) [inline addons/jr/fnc_compatibleItems.sqf]: `configProperties`,  `isClass`

File: addons/jr/fnc_compatibleItems.sqf [hunk 6]
```diff
@@ -15,32 +16,46 @@
...
+        ["'%1' not found in CfgWeapons",_weapon] call bis_fnc_error;
+    };
+};
 
+if (_typefilter == 0) then { //return
+    _compatibleItems
 } else {
-    ["'%1' not found in CfgWeapons",_weapon] call bis_fnc_error;
-    []
+    _compatibleItems select {_typefilter == getNumber(configfile>>"cfgweapons">>_x>>"itemInfo">>"type")};
```
Comment (Killswitch00) [inline addons/jr/fnc_compatibleItems.sqf]: `configFile`, `CfgWeapons`

**Comment by commy2** [comment]
This has one drawback. createLocation crashes the game when no mission is there. Really minor, but we don't know all the ways this function could be used.

**Comment by Killswitch00** [comment]
Crashing the game doesn't sound minor to me. Please elaborate on "when no mission is there" and when/how the crash happens. (Potential for MP abuse?)

**Comment by commy2** [comment]
For example when it's used in the main menu with `-world="empty"` or in a `CfgFunctions` with `preStart = 1;`
I really really doubt any addon out there does that, but if there would be, it would crash after this PR.
No MP abuse here.

**Comment by PabstMirror** [comment]
Could leave first function untouched and add a 2nd that does caching.
or `if (missionName == "") exitWith {[]};`

**Comment by commy2** [comment]
That might work, but again. This issue will most likely never happen.

**Comment by commy2** [comment]
I was wrong about the crash in the main menu. Can't reproduce anymore. This PR should work.

**Comment by Killswitch00** [comment]
Fix the comments so we can merge this.

**Pull request merged**